Badges Functionality + badge overview

bkatler opened the pull request:
In this PR:

- Added Badges for the following achievements: Creating 1, 5, or 10 stacks, Getting 5, 25, or 100 points, and Getting a 5, 15, or 30 day streak.
- When a user gets a badge, a pop-up comes on their screen which shows the badge and explains it
- Users can only get badges once
- To view badges, they hit the Badge Gallery Overview button on Profile (Profile is a slightly old version, will merge main back into here once the profile pr is approved)
- Badge Gallery Overview allows users to see all badges they have
- I added an animation which transitions across badges

I think it makes more sense to do profile pr first especially because it's only a few minor changes. When that happens I'll merge main in here and fix any conflicts asap :)

laurenindira commented on StudyStacks/Model/User.swift:
```diff
@@ -34,4 +34,6 @@ struct User: Identifiable, Codable {
     var favoriteStackIDs: [String]
 
+    // BADGES
+    var earnedBadges: [String] = []
```
don't assign values in the model itself

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -4,9 +4,10 @@
 //
 //  Created by Raihana Zahra on 3/8/25.
 //
-// medium.com/@jaredcassoutt/creating-tinder-like-swipeable-cards-in-swiftui-193fab1427b8
```
why are we taking out credits to sources...

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -4,9 +4,10 @@
-// medium.com/@jaredcassoutt/creating-tinder-like-swipeable-cards-in-swiftui-193fab1427b8
 
 import SwiftUI
+import Firebase
```
the reason why we have ViewModels is so that we don't do things like this. Firebase or any other service should not be referenced in the Views, this is why we encapsulate logic

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -142,16 +207,14 @@ struct CardStackView: View {
 
                 HStack {
-                    Button(action: {
-                        // TODO: Add thumbs-down action, save to firebase
```
why are you removing TODOs that aren't done

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -142,16 +207,14 @@ struct CardStackView: View {
+
             Spacer()
-            
-            //remember it section
```
don't remove comments that explain code. it's why we have them

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -163,9 +226,7 @@ struct CardStackView: View {
                         .foregroundColor(.black)
                         .padding(.horizontal)
 
-                    Button(action: {
```
same as above. don't remove undone TODOs

laurenindira commented on StudyStacks/View/Card/CardStackView.swift:
```diff
@@ -176,28 +237,53 @@ struct CardStackView: View {
     }
-}
 
-#Preview {
```
why would you get rid of a preview?????

laurenindira commented on StudyStacks/View/ContentView.swift:
```diff
@@ -21,34 +21,55 @@ struct ContentView: View {
                         .environmentObject(friendVM)
                         .tabItem {
-                            Label("Dashboard", systemImage: "house")
+                            Label("Dashboard", systemImage: "rectangle.stack.fill")
```
don't change the icons.

laurenindira commented on StudyStacks/View/ContentView.swift:
```diff
@@ -21,34 +21,55 @@ struct ContentView: View {
             }
         }
     }
+
```
remove this, it won't be used I assume. if we need it, we can look at the commit history

laurenindira commented on StudyStacks/View/ProfileView.swift:
see feedback from Profile PR

laurenindira commented on StudyStacks/View/Stacks/NewStackView.swift:
```diff
@@ -23,140 +23,107 @@ struct NewStackView: View {
-                                }
+                                .background(RoundedRectangle(cornerRadius: 15).fill(Color.surface))
                         }
-                        //TODO: add tags as dropdown instead of list
```
don't remove TODOS

laurenindira commented on StudyStacks/View/Stacks/NewStackView.swift:
```diff
@@ -23,140 +23,107 @@ struct NewStackView: View {
-                                    RoundedRectangle(cornerRadius: 15)
-                                        .fill(Color.surface)
-                                }
+                                .background(RoundedRectangle(cornerRadius: 15).fill(Color.surface))
```
this is now harder to read. why would you change it from what it was before

laurenindira commented on StudyStacks/View/Stacks/NewStackView.swift:
```diff
@@ -23,140 +23,107 @@ struct NewStackView: View {
+            ScrollView {
                 VStack {
                     VStack(alignment: .leading, spacing: 10) {
-                        //TITLE
```
stop getting rid of the comments!

laurenindira commented on StudyStacks/View/Stacks/NewStackView.swift:
lots to fix here. i don't know how you're writing this code, but there's no reason for there to be this many mistakes for what should have just been two functions added.

laurenindira commented on StudyStacks/ViewModel/AuthViewModel.swift:
```diff
@@ -12,6 +12,7 @@ import FirebaseFirestore
 import GoogleSignIn
 
 @Observable
+
```
no.

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
         self.isLoading = true
-        
         guard let _ = AuthViewModel.shared.user else {
-            print("CURRENTLY KNOWN USER FOR STACKS: \(String(describing: AuthViewModel.shared.user))")
```
don't delete debug print statements

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
+
     var creatingStack: Bool = false
     var editingStack: Bool = false
-    
```
separation was there for a reason

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
+
     private let db = Firestore.firestore()
-    
-    //delaying firebase writes for favorites
```
stop removing comments

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
             let querySnapshot = try await db.collection("allStacks").document(userID).collection("stacks").getDocuments()
             let stacks = querySnapshot.documents.compactMap { try? $0.data(as: Stack.self) }
             self.userStacks = stacks
-        } catch let error as NSError {
```
do you even know why you're changing this

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
+
         do {
-            try await stackRef.document(stack.id).delete()
+            try await db.collection("allStacks").document(user.id).collection("stacks").document(stack.id).delete()
```
this, in fact, reduces clarity.

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
-    
-    //MARK: - Favorite Stacks
+
+    // MARK: - Favorites
```
why'd you change the mark title

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
-            return
-        }
-        
+        guard let user = AuthViewModel.shared.user else { return }
```
why'd you change it to this?

laurenindira commented on StudyStacks/ViewModel/StackViewModel.swift:
```diff
@@ -17,188 +17,194 @@ class StackViewModel: ObservableObject {
-        print("SUCCESS: Cleared favorite data locally")
     }
-    
-    //TODO: revisit these later on in order to delay the sync for the favorites
```
why would you delete code that is still clearly marked as something to be used later

laurenindira reviewed (changes requested):
immediately obvious issues addressed, i need to test more on my end to comment on the logic. however, there was enough to submit an initial review

bkatler commented:
Resolved all issues, full notes:
- Restored all accidentally deleted credits, comments, and TODOs
- Removed default value of earnedBadges from User Model
- Removed all logic requiring firebase and Firestore inside CardStackView and created a helper function called updateStreakAndCheckForBadge() inside AuthViewModel
- Removed extra custom button from nav on ContentView
- Some other minor changes, mostly just small changes
